common/src/main/java/io/netty/util/concurrent/Promise.java

```diff
@@ -15,67 +15,99 @@
  */
 package io.netty.util.concurrent;
 
+import java.util.concurrent.CancellationException;
+
 /**
  * Special {@link Future} which is writable.
  */
-public interface Promise<V> extends Future<V> {
```

Comment:
Can we not bite this off in this PR? It seemed like the PR is changing enough as is, and this seems fundamentally a separate change.

It isn't clear to me how much value this adds, but I would _really_ appreciate that it'd make it harder to add listeners to promises which causes weird listener orderings. I don't _dislike_ it, but I have trouble identifying a way it provides enough value to justify the API migration.